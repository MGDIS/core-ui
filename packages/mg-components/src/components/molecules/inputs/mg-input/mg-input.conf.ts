/**
 * Possible input width
 */
export const widths = [2, 4, 16, 'full'] as const;

export type Width = (typeof widths)[number];

/**
 * Possible tootltip postions
 */
export const tooltipPositions = ['input', 'label'] as const;

export type TooltipPosition = (typeof tooltipPositions)[number];

/**
 * Possible label heading level
 */
export const labelHeadingLevels = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6'] as const;

export type LabelHeadingLevel = (typeof labelHeadingLevels)[number];

/**
 * Base interface for common methods
 */
interface InputBaseMethods {
  required: boolean;
  invalid: boolean;
  labelOnTop: boolean;
  readonly: boolean;
  displayError: () => Promise<void>;
  reset: () => Promise<void>;
}

/**
 * Possible Input types
 */
export type HTMLMgInputsElement =
  | (HTMLMgInputElement & InputBaseMethods)
  | HTMLMgInputCheckboxElement
  | HTMLMgInputDateElement
  | HTMLMgInputNumericElement
  | HTMLMgInputPasswordElement
  | HTMLMgInputRadioElement
  | HTMLMgInputSelectElement
  | HTMLMgInputTextElement
  | HTMLMgInputTextareaElement
  | (HTMLMgInputToggleElement & InputBaseMethods);

/**
 * Available WindowEvent types
 */
export type EventType = keyof WindowEventMap & 'blur';

export const classFieldset = 'mg-c-input--fieldset';
export const classDisabled = 'mg-c-input--disabled';
export const classReadonly = 'mg-c-input--readonly';
export const classVerticalList = 'mg-c-input--vertical-list';
