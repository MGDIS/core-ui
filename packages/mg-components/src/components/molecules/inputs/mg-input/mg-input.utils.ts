import { isValidString } from '@mgdis/core-ui-helpers/dist/utils';
import type { ErrorMessageDetailsType } from './mg-input.conf';

/**
 * Define error message
 * @param message - message to display
 * @param details - message detail to display
 * @returns formated MgInput["errorMessage"]
 */
export const defineErrorMessage = (message: string, details?: ErrorMessageDetailsType['details']): string | ErrorMessageDetailsType | undefined => {
  if (isValidString(message) && isValidString(details)) {
    return {
      summary: message,
      details,
    };
  } else if (isValidString(message)) {
    return message;
  } else {
    return undefined;
  }
};
