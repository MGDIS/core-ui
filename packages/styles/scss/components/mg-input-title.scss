@import 'form';

legend.mg-input-title {
  padding-right: 0;
  padding-left: 0;
}

.mg-input-title__required {
  display: var(--mg-input-title-display, unset);
}
