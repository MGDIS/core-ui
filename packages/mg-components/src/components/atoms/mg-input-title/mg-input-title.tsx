import { Component, h, Prop, State, Watch } from '@stencil/core';
import { isValidString } from '../../../utils/components.utils';

@Component({
  tag: 'mg-input-title',
  styleUrl: '../../../../node_modules/@mgdis/styles/dist/components/mg-input-title.css',
  scoped: true,
})
export class MgInputTitle {
  /**
   * Label input id
   */
  @Prop() identifier!: string;
  @Watch('identifier')
  validateIdentifier(newValue: MgInputTitle['identifier']): void {
    if (!isValidString(newValue)) {
      throw new Error('<mg-input-title> prop "identifier" is required.');
    }
  }

  /**
   * Define `legend|label` element id
   */
  @Prop() titleId: string;

  /**
   * If input is required an asterisk is added at the end of the label
   */
  @Prop() required: boolean;

  /**
   * If input is required an asterisk is added at the end of the label
   */
  @Prop() readonly: boolean;

  /**
   * Switch from label to fieldset sementic
   */
  @Prop() isLegend = false;
  @Watch('isLegend')
  validateIsLegend(newValue: MgInputTitle['isLegend']) {
    this.tagName = newValue ? 'legend' : 'label';
  }

  /**
   * Component parent tagname
   */
  @State() tagName = 'label';

  /*************
   * Lifecycle *
   *************/

  /**
   * Init tag name
   */
  componentWillLoad(): void {
    this.validateIdentifier(this.identifier);
    this.validateIsLegend(this.isLegend);
  }

  /**
   * Render
   * @returns HTML element
   */
  render(): HTMLElement {
    const TagName = this.readonly ? 'span' : this.tagName;
    // \u00A0 represent a &nbsp;
    return (
      <TagName class="mg-c-input-title" htmlFor={this.isLegend ? undefined : this.identifier} id={this.titleId}>
        <slot></slot>
        {this.required && (
          <span class="mg-c-input-title__required">
            &nbsp;<span class="mg-u-is-asterisk">*</span>
          </span>
        )}
      </TagName>
    );
  }
}
