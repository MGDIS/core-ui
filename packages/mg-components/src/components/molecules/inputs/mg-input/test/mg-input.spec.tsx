import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgInput } from '../../mg-input/mg-input';
import { MgInputTitle } from '../../../../atoms/mg-input-title/mg-input-title';
import { classDisabled, classFieldset, classReadonly, classVerticalList } from '../mg-input.conf';

const baseArgs = {
  label: 'label',
  identifier: 'identifier',
};

const errorMessage = 'This is an error';

const helpText = 'Hello joker';

const tooltip = 'Batman is a DC Comics license';

const getPage = (args = {}, slot?) => {
  const props = { ...baseArgs, ...args };

  return newSpecPage({
    components: [MgInputTitle, MgInput],
    template: () => <mg-input {...props}>{slot ? slot : <input type="file" id={props.identifier}></input>}</mg-input>,
  });
};

describe('mg-input', () => {
  describe('render', () => {
    test.each([
      {},
      { class: '' },
      { labelOnTop: true },
      { labelHide: true },
      { required: true },
      { helpText },
      { errorMessage },
      // fieldset
      { class: classFieldset },
      // readonly
      { class: classReadonly, readonlyValue: 'batman' },
      { class: classReadonly, readonlyValue: ['batman'] },
      { class: classReadonly, readonlyValue: ['batman', 'joker', 'bane'] },
      { class: [classReadonly, classVerticalList].join(' '), readonlyValue: 'batman' },
      { class: [classReadonly, classVerticalList].join(' '), readonlyValue: ['batman', 'joker', 'bane'] },
      { class: [classReadonly, classVerticalList].join(' '), readonlyValue: ['batman'] },
      { class: classReadonly, errorMessage },
      { class: classReadonly },
      { class: classReadonly, inputVerticalList: true },
      { class: classReadonly, required: true, helpText },
      { class: classReadonly, labelOnTop: true, tooltip },
      // disabled
      { class: classDisabled },
      { class: classDisabled, required: true, helpText },
      { class: classDisabled, errorMessage },
      // tooltip
      { tooltip },
      { tooltip, tooltipPosition: 'label' },
      { tooltip, tooltipPosition: 'input', labelOnTop: true },
    ])('Should render with args %s:', async args => {
      const { root } = await getPage(args);
      expect(root).toMatchSnapshot();
    });

    test('Should update "label" slot', async () => {
      const page = await getPage({ label: 'batma,' });
      expect(page).toMatchSnapshot();
      const element = page.doc.querySelector('mg-input');

      element.label = 'joker';
      await page.waitForChanges();

      expect(page.root).toMatchSnapshot();
    });

    test('Should update "error" slot', async () => {
      const page = await getPage();
      const element = page.doc.querySelector('mg-input');

      element.errorMessage = errorMessage;
      await page.waitForChanges();

      expect(page.root).toMatchSnapshot();

      element.errorMessage = undefined;
      await page.waitForChanges();

      expect(page.root).toMatchSnapshot();
    });

    test('Should update "help-text" slot', async () => {
      const page = await getPage();
      const element = page.doc.querySelector('mg-input');

      element.helpText = helpText;
      await page.waitForChanges();

      expect(page.root).toMatchSnapshot();

      element.helpText = undefined;
      await page.waitForChanges();

      expect(page.root).toMatchSnapshot();
    });
  });

  describe('errors', () => {
    test.each(['', ' ', undefined])('Should not render with invalid "identifier" property: %s', async identifier => {
      expect.assertions(1);
      try {
        await getPage({ identifier });
      } catch (err) {
        expect(err.message).toMatch('<mg-input> prop "identifier" is required.');
      }
    });

    test.each(['', ' ', undefined])('Should not render with invalid "label" property: %s', async label => {
      expect.assertions(1);
      try {
        await getPage({ label });
      } catch (err) {
        expect(err.message).toMatch('<mg-input> prop "label" is required.');
      }
    });

    test.each([' ', 'batman'])('Should not render with invalid "tooltipPosition" property: %s', async tooltipPosition => {
      expect.assertions(1);
      try {
        await getPage({ tooltipPosition });
      } catch (err) {
        expect(err.message).toMatch('<mg-input> prop "tooltipPosition" must be one of: input, label');
      }
    });

    test('Should not render with label-on-top and label-hide setup', async () => {
      expect.assertions(1);
      try {
        await getPage({ ...baseArgs, labelOnTop: true, labelHide: true });
      } catch (err) {
        expect(err.message).toMatch('<mg-input> prop "labelOnTop" must not be paired with the prop "labelHide".');
      }
    });

    test('Should throw error when slotted input hasn\'t valid "id" attribute', async () => {
      expect.assertions(1);
      try {
        await getPage({}, <input type="file"></input>);
      } catch (err) {
        expect(err.message).toMatch('<mg-input> "identifier" prop has no target for id: identifier. Add an id to the targeted input.');
      }
    });
  });
});
