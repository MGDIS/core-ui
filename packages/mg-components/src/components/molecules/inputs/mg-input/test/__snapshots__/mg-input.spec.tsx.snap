// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-input render Should render with args {
  classCollection: [xe],
  labelOnTop: true,
  tooltip: 'Batman is a DC Comics license'
}: 1`] = `
<mg-input class="mg-c-input mg-c-input--label-on-top mg-c-input--readonly">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <strong></strong>
      <slot></slot>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="true" required="false" slot="label">
    <!---->
    <span class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </span>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args {
  tooltip: 'Batman is a DC Comics license',
  tooltipPosition: 'input',
  labelOnTop: true
}: 1`] = `
<mg-input class="mg-c-input mg-c-input--label-on-top">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
      <mg-tooltip identifier="identifier-tooltip" message="Batman is a DC Comics license">
        <mg-icon icon="info-circle"></mg-icon>
      </mg-tooltip>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { classCollection: [xe] }: 1`] = `
<mg-input class="mg-c-input">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { classCollection: [xe] }: 2`] = `
<mg-input aria-labelledby="identifier-title" class="mg-c-input mg-c-input--fieldset" role="group">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="true" readonly="false" required="false" slot="label">
    <!---->
    <legend class="mg-c-input-title" id="identifier-title">
      <span class="mg-c-input-title__text">
        label
      </span>
    </legend>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { classCollection: [xe] }: 3`] = `
<mg-input class="mg-c-input mg-c-input--readonly">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <strong></strong>
      <slot></slot>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="true" required="false" slot="label">
    <!---->
    <span class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </span>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { classCollection: [xe] }: 4`] = `
<mg-input class="mg-c-input mg-c-input--disabled">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { classCollection: [xe], inputVerticalList: true }: 1`] = `
<mg-input class="mg-c-input mg-c-input--readonly" inputverticallist="">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <strong></strong>
      <slot></slot>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="true" required="false" slot="label">
    <!---->
    <span class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </span>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { errorMessage: 'This is an error' }: 1`] = `
<mg-input class="mg-c-input mg-c-input--has-error">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
      <div aria-live="assertive" class="mg-c-input__error">
        <slot name="error"></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="identifier-error" id="identifier" type="file">
  <div id="identifier-error" slot="error">
    This is an error
  </div>
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { errorMessage: 'This is an error', classCollection: [xe] }: 1`] = `
<mg-input class="mg-c-input mg-c-input--has-error mg-c-input--readonly">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <strong></strong>
      <slot></slot>
    </div>
  </mock:shadow-root>
  <input aria-describedby="identifier-error" id="identifier" type="file">
  <div id="identifier-error" slot="error">
    This is an error
  </div>
  <mg-input-title identifier="identifier" is-legend="false" readonly="true" required="false" slot="label">
    <!---->
    <span class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </span>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { errorMessage: 'This is an error', classCollection: [xe] }: 2`] = `
<mg-input class="mg-c-input mg-c-input--disabled mg-c-input--has-error">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="identifier-error" id="identifier" type="file">
  <div id="identifier-error" slot="error">
    This is an error
  </div>
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { helpText: 'Hello joker' }: 1`] = `
<mg-input class="mg-c-input">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
      <div class="mg-c-input__help-text">
        <slot name="help-text"></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="identifier-help-text" id="identifier" type="file">
  <div id="identifier-help-text" slot="help-text">
    Hello joker
  </div>
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { labelHide: true }: 1`] = `
<mg-input class="mg-c-input">
  <mock:shadow-root>
    <div class="mg-c-input__title mg-u-visually-hidden">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { labelOnTop: true }: 1`] = `
<mg-input class="mg-c-input mg-c-input--label-on-top">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { readonlyValue: 'batman', classCollection: [xe] }: 1`] = `
<mg-input class="mg-c-input mg-c-input--readonly">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <strong>
        batman
      </strong>
      <slot></slot>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="true" required="false" slot="label">
    <!---->
    <span class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </span>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { readonlyValue: [Array], classCollection: [xe] }: 1`] = `
<mg-input class="mg-c-input mg-c-input--readonly">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <ul>
        <li>
          <strong>
            batman
          </strong>
        </li>
        <li>
          <strong>
            joker
          </strong>
        </li>
        <li>
          <strong>
            bane
          </strong>
        </li>
      </ul>
      <slot></slot>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="true" required="false" slot="label">
    <!---->
    <span class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </span>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { required: true }: 1`] = `
<mg-input class="mg-c-input">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="true" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
      <span class="mg-c-input-title__required">
        <span class="mg-u-is-asterisk">
          *
        </span>
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { required: true, classCollection: [xe], helpText: 'Hello joker' }: 1`] = `
<mg-input class="mg-c-input mg-c-input--readonly">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <strong></strong>
      <slot></slot>
    </div>
  </mock:shadow-root>
  <input aria-describedby="identifier-help-text" id="identifier" type="file">
  <div id="identifier-help-text" slot="help-text">
    Hello joker
  </div>
  <mg-input-title identifier="identifier" is-legend="false" readonly="true" required="true" slot="label">
    <!---->
    <span class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </span>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { required: true, classCollection: [xe], helpText: 'Hello joker' }: 2`] = `
<mg-input class="mg-c-input mg-c-input--disabled">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
      <div class="mg-c-input__help-text">
        <slot name="help-text"></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="identifier-help-text" id="identifier" type="file">
  <div id="identifier-help-text" slot="help-text">
    Hello joker
  </div>
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { tooltip: 'Batman is a DC Comics license' }: 1`] = `
<mg-input class="mg-c-input">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
        <mg-tooltip identifier="identifier-tooltip" message="Batman is a DC Comics license">
          <mg-icon icon="info-circle"></mg-icon>
        </mg-tooltip>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args { tooltip: 'Batman is a DC Comics license', tooltipPosition: 'label' }: 1`] = `
<mg-input class="mg-c-input">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
      <mg-tooltip identifier="identifier-tooltip" message="Batman is a DC Comics license">
        <mg-icon icon="info-circle"></mg-icon>
      </mg-tooltip>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should render with args {}: 1`] = `
<mg-input class="mg-c-input">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should update "error" slot 1`] = `
<mg-input class="mg-c-input mg-c-input--has-error">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
      <div aria-live="assertive" class="mg-c-input__error">
        <slot name="error"></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="identifier-error" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
  <div id="identifier-error" slot="error">
    This is an error
  </div>
</mg-input>
`;

exports[`mg-input render Should update "error" slot 2`] = `
<mg-input class="mg-c-input">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should update "help-text" slot 1`] = `
<mg-input class="mg-c-input">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
      <div class="mg-c-input__help-text">
        <slot name="help-text"></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="identifier-help-text" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
  <div id="identifier-help-text" slot="help-text">
    Hello joker
  </div>
</mg-input>
`;

exports[`mg-input render Should update "help-text" slot 2`] = `
<mg-input class="mg-c-input">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        label
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;

exports[`mg-input render Should update "label" slot 1`] = `
{
  "body": <body>
  <mg-input class="mg-c-input">
    <mock:shadow-root>
      <div class="mg-c-input__title">
        <slot name="label"></slot>
      </div>
      <div class="mg-c-input__input-container">
        <div class="mg-c-input__input">
          <slot></slot>
        </div>
      </div>
    </mock:shadow-root>
    <input aria-describedby="" id="identifier" type="file">
    <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
      <!---->
      <label class="mg-c-input-title" htmlfor="identifier">
        <span class="mg-c-input-title__text">
          batma,
        </span>
      </label>
    </mg-input-title>
  </mg-input>
</body>,
  "build": {
    "allRenderFn": true,
    "appendChildSlotFix": false,
    "asyncLoading": true,
    "asyncQueue": true,
    "attachStyles": false,
    "cloneNodeFix": false,
    "cmpDidLoad": true,
    "cmpDidRender": true,
    "cmpDidUnload": true,
    "cmpDidUpdate": true,
    "cmpShouldUpdate": true,
    "cmpWillLoad": true,
    "cmpWillRender": true,
    "cmpWillUpdate": true,
    "connectedCallback": true,
    "constructableCSS": true,
    "cssAnnotations": false,
    "devTools": false,
    "disconnectedCallback": true,
    "element": true,
    "event": true,
    "experimentalScopedSlotChanges": false,
    "experimentalSlotFixes": false,
    "formAssociated": true,
    "hasRenderFn": true,
    "hostListener": true,
    "hostListenerTarget": true,
    "hostListenerTargetBody": true,
    "hostListenerTargetDocument": true,
    "hostListenerTargetParent": true,
    "hostListenerTargetWindow": true,
    "hotModuleReplacement": false,
    "hydrateClientSide": false,
    "hydrateServerSide": false,
    "hydratedAttribute": false,
    "hydratedClass": true,
    "initializeNextTick": true,
    "invisiblePrehydration": true,
    "isDebug": false,
    "isDev": true,
    "isTesting": true,
    "lazyLoad": true,
    "lifecycle": true,
    "lifecycleDOMEvents": true,
    "member": true,
    "method": true,
    "mode": true,
    "observeAttribute": true,
    "profile": true,
    "prop": true,
    "propBoolean": true,
    "propMutable": true,
    "propNumber": true,
    "propString": true,
    "reflect": true,
    "scoped": true,
    "scopedSlotTextContentFix": false,
    "scriptDataOpts": false,
    "shadowDelegatesFocus": true,
    "shadowDom": true,
    "shadowDomShim": false,
    "slot": true,
    "slotChildNodesFix": false,
    "slotRelocation": true,
    "state": true,
    "style": true,
    "svg": true,
    "taskQueue": true,
    "transformTagName": true,
    "updatable": true,
    "vdomAttribute": true,
    "vdomClass": true,
    "vdomFunctional": true,
    "vdomKey": true,
    "vdomListener": true,
    "vdomPropOrAttr": true,
    "vdomRef": true,
    "vdomRender": true,
    "vdomStyle": true,
    "vdomText": true,
    "vdomXlink": true,
    "watchCallback": true,
  },
  "doc": ,
  "flushLoadModule": [Function],
  "flushQueue": [Function],
  "setContent": [Function],
  "styles": Map {},
  "waitForChanges": [Function],
  "win": [MockWindow],
}
`;

exports[`mg-input render Should update "label" slot 2`] = `
<mg-input class="mg-c-input">
  <mock:shadow-root>
    <div class="mg-c-input__title">
      <slot name="label"></slot>
    </div>
    <div class="mg-c-input__input-container">
      <div class="mg-c-input__input">
        <slot></slot>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="" id="identifier" type="file">
  <mg-input-title identifier="identifier" is-legend="false" readonly="false" required="false" slot="label">
    <!---->
    <label class="mg-c-input-title" htmlfor="identifier">
      <span class="mg-c-input-title__text">
        joker
      </span>
    </label>
  </mg-input-title>
</mg-input>
`;
